File: python/src/pi_mono/utils/open_browser.py
import sys
import subprocess


def open_browser(target: str) -> None:
    """Open a URL or file in the platform browser/default handler without invoking a shell."""
    if sys.platform == "darwin":
        cmd, args = "open", [target]
    elif sys.platform == "win32":
        cmd, args = "rundll32", ["url.dll,FileProtocolHandler", target]
    else:
        cmd, args = "xdg-open", [target]

    try:
        subprocess.Popen(
            [cmd] + args,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            stdin=subprocess.DEVNULL,
            close_fds=True,
        )
    except Exception:
        pass

File: packages/coding-agent/src/utils/open-browser.ts
import { spawn } from "node:child_process";
import { join } from "node:path";

/**
 * Open a URL or file in the platform browser/default handler.
 *
 * This intentionally never invokes a shell. On Windows, do not use
 * `cmd /c start`: cmd.exe re-parses metacharacters (&, |, ^, ...) before
 * `start` runs, which would make attacker-controlled URLs injectable.
 */
export function openBrowser(target: string): void {
	const [cmd, args]: [string, string[]] =
		process.platform === "darwin"
			? ["open", [target]]
			: process.platform === "win32"
				? [
						join(process.env.SystemRoot ?? "C:\\Windows", "System32", "rundll32.exe"),
						["url.dll,FileProtocolHandler", target],
					]
				: ["xdg-open", [target]];

	// spawn reports launcher failures (for example, missing xdg-open) via an
	// error event. Browser launch is best-effort: callers still present the target
	// to the user, so keep the launcher failure from becoming a process crash.
	spawn(cmd, args, { stdio: "ignore", detached: true })
		.on("error", () => {})
		.unref();
}
